tools/ci-cdk/canary-runner/src/build_bundle.rs

```diff
@@ -0,0 +1,462 @@
+use std::fmt::Write as FmtWrite;
+use std::fs;
+use std::io::Write as IoWrite;
+use std::path::{Path, PathBuf};
+use std::process::Command;
+use std::str::FromStr;
+
+const BASE_MANIFEST: &str = r#"
+# Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+# SPDX-License-Identifier: Apache-2.0
+#
+# IMPORTANT: Don't edit this file directly! Run `canary-runner build-bundle` to modify this file instead.
+[package]
+name = "aws-sdk-rust-lambda-canary"
+version = "0.1.0"
+edition = "2021"
+license = "Apache-2.0"
+
+# Emit an empty workspace so that the canary can successfully build when
+# built from the aws-sdk-rust repo, which has a workspace in it.
+[workspace]
+
+[[bin]]
+name = "bootstrap"
+path = "src/main.rs"
+
+[dependencies]
+anyhow = "1"
+async-stream = "0.3"
+bytes = "1"
+hound = "3.4"
+async-trait = "0.1"
+lambda_runtime = "0.4"
+serde_json = "1"
+thiserror = "1"
+tokio = { version = "1", features = ["full"] }
+tracing = "0.1"
+tracing-subscriber = { version = "0.3", features = ["fmt", "env-filter"] }
+uuid = { version = "0.8", features = ["v4"] }
+tokio-stream = "0"
+tracing-texray = "0.1.1"
+"#;
+
+const REQUIRED_SDK_CRATES: &[&str] = &[
+    "aws-config",
+    "aws-sdk-s3",
+    "aws-sdk-ec2",
+    "aws-sdk-transcribestreaming",
+];
+
+lazy_static! {
+    static ref NOTABLE_SDK_RELEASE_TAGS: Vec<ReleaseTag> = vec![
+        ReleaseTag::from_str("v0.4.1").unwrap(), // first version to add support for paginators
+    ];
+}
+
+#[derive(Debug, Parser, Eq, PartialEq)]
+pub struct BuildBundleOpt {
+    /// Canary Lambda source code path (defaults to current directory)
+    #[clap(long)]
+    pub canary_path: Option<PathBuf>,
+
+    /// SDK release tag to use for crate version numbers
+    #[clap(
+        long,
+        required_unless_present = "sdk-path",
+        conflicts_with = "sdk-path"
+    )]
+    pub sdk_release_tag: Option<ReleaseTag>,
+
+    /// SDK release tag to use for crate version numbers
+    #[clap(
+        long,
+        required_unless_present = "sdk-release-tag",
+        conflicts_with = "sdk-release-tag"
+    )]
+    pub sdk_path: Option<PathBuf>,
+
+    /// Whether to target MUSL instead of GLIBC when compiling the Lambda
+    #[clap(long)]
+    pub musl: bool,
+
+    /// Only generate the `Cargo.toml` file rather than building the entire bundle
+    #[clap(long)]
+    pub manifest_only: bool,
+}
+
+enum CrateSource {
+    Path(PathBuf),
+    VersionsManifest {
+        versions: VersionsManifest,
+        release_tag: ReleaseTag,
+    },
+}
+
+fn enabled_features(crate_source: &CrateSource) -> Vec<String> {
+    let mut enabled = Vec::new();
+    if let CrateSource::VersionsManifest { release_tag, .. } = crate_source {
+        for notable in NOTABLE_SDK_RELEASE_TAGS.iter() {
+            if notable < release_tag {
+                enabled.push(notable.as_str().into());
+            }
+        }
+    }
+    enabled
+}
+
+fn generate_crate_manifest(crate_source: CrateSource) -> Result<String> {
```

Comment:
It is perhaps overkill, but we could use cargo's stuff for this in the future https://docs.rs/cargo/0.63.1/cargo/core/manifest/struct.Manifest.html

Reply:
Depending on `cargo` would definitely be overkill, but there are some smaller crates that just operate on the Cargo manifest file. The one I've used before ends up serializing all the optional properties that aren't set though, so I'm hesitant to go that route.